src/functions/messages/put_function.rs

```diff
@@ -0,0 +1,130 @@
+use std::{collections::HashMap, iter::FromIterator, time::Duration};
+
+use crate::{
+    functions::{EnvironmentValue, Function, FunctionClient, MomentoRequest, WasmSource},
+    utils::prep_request_with_timeout,
+    MomentoError, MomentoResult,
+};
+
+/// Create or update a Function.
+/// The cache is used as a namespace for your Functions.
+///
+/// # Arguments
+///
+/// * `cache_name` - The name of the cache to use as a namespace for the Function.
+/// * `function_name` - The name of the Function.
+///
+/// # Example
+///
+/// ```rust
+/// # fn main() -> anyhow::Result<()> {
+/// # tokio_test::block_on(async {
+/// use momento::{CredentialProvider, FunctionClient};
+/// use momento::functions::PutFunctionRequest;
+/// # use momento_test_util::echo_wasm;
+/// # let (function_client, cache_name) = momento_test_util::create_doctest_function_client();
+/// // load your wasm from a .wasm file compiled with wasm32-wasip2
+/// let function_body = echo_wasm();
+///
+/// let request = PutFunctionRequest::new(cache_name, "hello functions", function_body);
+/// let function = function_client.send(request).await?;
+/// println!("Created a function: {function:?}");
+/// # Ok(())
+/// # })
+/// # }
+/// ```
+pub struct PutFunctionRequest {
+    cache_name: String,
+    function_name: String,
+    description: String,
+    environment: HashMap<String, EnvironmentValue>,
+    wasm_source: WasmSource,
+}
+
+impl PutFunctionRequest {
+    /// Create a new PublishRequest.
+    pub fn new(
+        cache_name: impl Into<String>,
+        function_name: impl Into<String>,
+        wasm_source: impl Into<WasmSource>,
+    ) -> Self {
+        Self {
+            cache_name: cache_name.into(),
+            function_name: function_name.into(),
+            description: String::new(),
+            environment: HashMap::new(),
+            wasm_source: wasm_source.into(),
+        }
+    }
+
+    /// Set the Function's description
+    pub fn with_description(mut self, description: impl Into<String>) -> Self {
+        self.description = description.into();
+        self
+    }
```

Comment:
Here's an example of [how we model](https://github.com/momentohq/client-sdk-rust/blob/main/src/cache/messages/data/sorted_set/sorted_set_fetch_by_score.rs#L66-L119) rich requests with defaults. Can apply that pattern here.

Reply:
Optional things are `Option`. That allows developers to pass in `None` (vs a default value) when instantiating the struct or using the setters. This was pattern was requested for ease of use with `rpc-perf`.

Setters are also named as just the name of field and accept an `Option`.